drivers/pio/sunxi-pio.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <bitmap.h>
+#include <debug.h>
+#include <dm.h>
+#include <error.h>
+#include <mmio.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <util.h>
+#include <drivers/clock.h>
+#include <drivers/irqchip.h>
+#include <drivers/msgbox.h>
+#include <drivers/msgbox/sunxi-msgbox.h>
+#include <drivers/pio.h>
+#include <drivers/pio/sunxi-pio.h>
+
+#define PL_CFG0_REG          0x00
+#define PL_CFG0_REG_CLR      0x77777777
+#define PL_CFG0_REG_NUM_PINS 7
+#define PL_CFG1_REG          0x04
+#define PL_CFG1_REG_CLR      0x00077777
+#define PL_DATA_REG          0x10
+#define PL_DATA_REG_OFFSET   4
+#define PL_EINT_CTL_REG      0x210
+#define PL_EINT_STA_REG      0x214
+
+static void
+sunxi_pio_irq(struct device *dev)
+{
+	/* Clear all processed pending interrupts. */
+	mmio_setbits32(dev->regs + PL_EINT_STA_REG, 1);
+}
+
+static int
+sunxi_pio_probe(struct device *dev)
+{
+	int err;
+
+	/* Disable IO for all ports. */
+	mmio_write32(dev->regs + PL_CFG0_REG, PL_CFG0_REG_CLR);
+	mmio_write32(dev->regs + PL_CFG1_REG, PL_CFG1_REG_CLR);
+
+	/* Disable and clear all IRQs. */
+	mmio_setbits32(dev->regs + PL_EINT_CTL_REG, BITMASK(0, 12));
+	mmio_setbits32(dev->regs + PL_EINT_STA_REG, BITMASK(0, 12));
+
+	if ((err = irqchip_register_irq(dev, sunxi_pio_irq)))
```

Comment:
Not handling IRQs for the moment, this can be dropped.

Reply:
Done